_posts/2018-11-29-GraphQL-Stitching.markdown

```diff
@@ -0,0 +1,386 @@
+---
+layout: epic
+title: GraphQL Stitching 101
+date: 2018-11-29
+author: [orta]
+categories: [graphql, stitching, metaphysics]
+css: graphql-stitch
+comment_id: 450
+---
+
+Micro-Services make sense. You can scope a domain of your business to particular small unit of abstraction like an
+API. Doing so makes it easy to work in isolation, experiment with new ideas and evolve in many directions.
+
+We've been [carefully pushing][monolith] for years to move away from our single monolithic API, to a collection of
+smaller, more focused projects. Our [highlights docs][high] showcase this well. The movement to smaller composable
+services works great from an isolated platform/systems perspective but can be a bit tricky to handle with front-end
+clients. Until 2018, the way that we've addressed the growing complexity in our service later has been to migrate
+the complexity inside our main GraphQL API, [metaphysics][mp]. Metaphysics is our GraphQL API gateway that
+consolidates many API sources into a single service, then extending and interleaving and their data to make clients
+easier to write.
+
+However, as more services have been created, and grown - so has metaphysics. This creates a worrying trend, as the
+growth of code in metaphysics isn't quite linear.
```

Comment:
The precise phrase I've been using lately to describe this is "super-linear" (above linear), but it's up to you how fancy mathy you want to be.

Reply:
Additionally, it _may_ be worth mentioning that Metaphysics acts both as an orchestration layer (coordinating API calls to various backends) and also an API in its own right (with its own business logic). Because it's both (neither?) we end up with a lot of complexity. Starting fresh today, I could see us _just_ treating metaphysics as an orchestrator through merging/stitching alone. It could be worth mentioning this tension up front – I know it's discussed near the end, but it might be a nice bookend. Up to you!

Reply:
Spent some time looking at this and I can't elegantly hit the beat I wanted to make (priming that we're already doing it) and include some of this extra info that we can also distinguish that there's extra bits that are MP only - gonna not change it, thanks though - it's a good point